crypto/x509/x_x509.c

```diff
@@ -244,3 +244,15 @@ int X509_get_signature_nid(const X509 *x)
 {
     return OBJ_obj2nid(x->sig_alg.algorithm);
 }
+
+#ifndef OPENSSL_NO_SM2
+void X509_set_sm2_id(X509 *x, ASN1_OCTET_STRING *sm2_id)
+{
+    x->sm2_id = *sm2_id;
```

Comment:
This is doing a shallow copy of the `ASN1_OCTET_STRING`. If `sm2_id` is allocated on the heap and then freed you'll be left with `x` storing a `ASN1_OCTET_STRING` with it's `data` pointer deallocated which will result in memory corruption.